File: test/test_solver.f90
! Tests for the module rbf_solver: the Fortran face of rbf_solver.h,
! exercised the way a Fortran driver would call it -- the generic
! solve_sparse with optional arguments present and absent, and the
! matrix-free form with a bind(c) callback and a context pointer. The
! system is the one-dimensional three-point Laplacian with Dirichlet
! boundary, symmetric positive definite, and the right-hand side is
! manufactured from a known solution.
!
! Built and run with the C++ test through CMake:
!
!   cmake -B build -DGPU_ASSEMBLY_ENABLE_SOLVER=ON
!   cmake --build build
!   ctest --test-dir build

! The matrix, its product, and the callbacks. Module procedures rather
! than internal ones: a bind(c) callback must not need a trampoline.
module test_solver_support
use, intrinsic :: iso_c_binding, only: c_int, c_double, c_ptr, c_f_pointer
use rbf_ellpack, only: ellpack_mv
implicit none
private

public :: csr, laplacian_1d, apply, matvec, identity
public :: ell, to_ellpack, ellpack_matvec
public :: check, failures

! In CSR, 0-based, handed to the callback through a pointer
type :: csr
    integer(c_int) :: n
    integer(c_int), allocatable :: ia(:), ja(:)
    real(c_double), allocatable :: val(:)
end type

! The same in ELLPACK, a(n, nnzrow), for the rbf_ellpack product
type :: ell
    integer(c_int) :: n, nnzrow
    integer(c_int), allocatable :: ja(:, :)
    real(c_double), allocatable :: a(:, :)
end type

integer :: failures = 0

contains

    subroutine check(cond, what)
        logical, intent(in) :: cond
        character(*), intent(in) :: what
        if (.not. cond) then
            print '(a,a)', "FAIL ", what
            failures = failures + 1
        end if
    end subroutine

    ! The three-point Laplacian -u'' with Dirichlet boundary
    subroutine laplacian_1d(n, A)
        integer, intent(in) :: n
        type(csr), intent(out) :: A
        integer :: i, p
        A%n = n
        allocate(A%ia(n + 1), A%ja(3*n - 2), A%val(3*n - 2))
        p = 0
        A%ia(1) = 0
        do i = 1, n
            if (i > 1) then
                p = p + 1
                A%ja(p) = i - 2
                A%val(p) = -1
            end if
            p = p + 1
            A%ja(p) = i - 1
            A%val(p) = 2
            if (i < n) then
                p = p + 1
                A%ja(p) = i
                A%val(p) = -1
            end if
            A%ia(i + 1) = p
        end do
    end subroutine

    ! y := y + alpha A x
    subroutine apply(A, alpha, x, y)
        type(csr), intent(in) :: A
        real(c_double), intent(in) :: alpha, x(:)
        real(c_double), intent(inout) :: y(:)
        integer :: i, p
        do i = 1, A%n
            do p = A%ia(i) + 1, A%ia(i + 1)
                y(i) = y(i) + alpha*A%val(p)*x(A%ja(p) + 1)
            end do
        end do
    end subroutine

    ! The product of the matrix behind data, of the interface matvec_t
    subroutine matvec(nr, nc, alpha, x, y, data) bind(c)
        integer(c_int), value :: nr, nc
        real(c_double), value :: alpha
        real(c_double), intent(in) :: x(nc)
        real(c_double), intent(inout) :: y(nr)
        type(c_ptr), value :: data
        type(csr), pointer :: A
        call c_f_pointer(data, A)
        call check(nr == A%n .and. nc == A%n, "matvec: dimensions")
        call apply(A, alpha, x, y)
    end subroutine

    ! The tridiagonal matrix as ELLPACK with three entries per row; the
    ! end rows repeat the diagonal index with a zero entry
    subroutine to_ellpack(A, E)
        type(csr), intent(in) :: A
        type(ell), intent(out) :: E
        integer :: i, j, p
        E%n = A%n
        E%nnzrow = 3
        allocate(E%ja(A%n, 3), E%a(A%n, 3))
        do i = 1, A%n
            E%ja(i, :) = i - 1
            E%a(i, :) = 0
            j = 0
            do p = A%ia(i) + 1, A%ia(i + 1)
                j = j + 1
                E%ja(i, j) = A%ja(p)
                E%a(i, j) = A%val(p)
            end do
        end do
    end subroutine

    ! rbf_ellpack's product as the operator of the matrix-free solver
    subroutine ellpack_matvec(nr, nc, alpha, x, y, data) bind(c)
        integer(c_int), value :: nr, nc
        real(c_double), value :: alpha
        real(c_double), intent(in) :: x(nc)
        real(c_double), intent(inout) :: y(nr)
        type(c_ptr), value :: data
        type(ell), pointer :: E
        call c_f_pointer(data, E)
        call check(nr == E%n .and. nc == E%n, "ellpack_matvec: dimensions")
        call ellpack_mv(E%n, E%nnzrow, alpha, E%a, E%ja, E%n, x, 1.0_c_double, y)
    end subroutine

    ! The identity, needing no context
    subroutine identity(nr, nc, alpha, x, y, data) bind(c)
        integer(c_int), value :: nr, nc
        real(c_double), value :: alpha
        real(c_double), intent(in) :: x(nc)
        real(c_double), intent(inout) :: y(nr)
        type(c_ptr), value :: data
        y = y + alpha*x
    end subroutine

end module

program test_solver
use, intrinsic :: iso_c_binding, only: c_int, c_double, c_loc, c_null_ptr
use rbf_solver
use test_solver_support
implicit none

integer, parameter :: n = 200
type(csr), target :: A
type(ell), target :: E
real(c_double) :: u(n), b(n), x(n), y(n), err
integer(c_int) :: status, iter
integer :: i

call laplacian_1d(n, A)
do i = 1, n
    u(i) = sin(0.05_c_double*i) + 0.01_c_double*i
end do
b = 0
call apply(A, 1.0_c_double, u, b)

! Every optional argument absent: BiCGSTAB with the Jacobi preconditioner
x = 0
status = solve_sparse(A%n, size(A%val), A%val, A%ia, A%ja, b, x)
call check(status == SOLVER_SUCCESS, "csr defaults: status")
call check(maxval(abs(x - u)) < 1.0e-8_c_double, "csr defaults: solution")

! Conjugate gradient with incomplete Cholesky, keywords, the results asked for
x = 0
status = solve_sparse(A%n, size(A%val), A%val, A%ia, A%ja, b, x, &
    res_error=err, res_iter=iter, method=SOLVER_CG, &
    precond=PRECOND_ILU, tolerance=1.0e-12_c_double)
call check(status == SOLVER_SUCCESS, "csr cg/ichol: status")
call check(err >= 0 .and. err <= 1.0e-12_c_double, "csr cg/ichol: residual")
! Eigen counts completed passes, so an exact preconditioner reports 0
call check(iter >= 0 .and. iter <= 2*n, "csr cg/ichol: iterations")
call check(maxval(abs(x - u)) < 1.0e-9_c_double, "csr cg/ichol: solution")

! Too few iterations: NO_CONVERGENCE, with the count reported
x = 0
status = solve_sparse(A%n, size(A%val), A%val, A%ia, A%ja, b, x, &
    res_iter=iter, max_iter=1, tolerance=1.0e-14_c_double)
call check(status == SOLVER_NO_CONVERGENCE, "csr limit: status")
call check(iter == 1, "csr limit: iterations")

! Matrix-free, the matrix reaching the callback through the context pointer
x = 0
status = solve_sparse(n, n, matvec, c_loc(A), b, x, res_error=err, res_iter=iter, &
    method=SOLVER_CG, tolerance=1.0e-12_c_double)
call check(status == SOLVER_SUCCESS, "mf cg: status")
call check(err >= 0 .and. err <= 1.0e-12_c_double, "mf cg: residual")
call check(maxval(abs(x - u)) < 1.0e-9_c_double, "mf cg: solution")

! Matrix-free with the ELLPACK product of rbf_ellpack as the operator
call to_ellpack(A, E)
x = 0
status = solve_sparse(n, n, ellpack_matvec, c_loc(E), b, x, res_error=err, &
    tolerance=1.0e-12_c_double)
call check(status == SOLVER_SUCCESS, "mf ellpack: status")
call check(maxval(abs(x - u)) < 1.0e-9_c_double, "mf ellpack: solution")

! Eigen's CSR product against the reference: y = A u, then y = -0.5 y + 2 A u
y = huge(1.0_c_double)
call csr_mv(A%n, A%n, size(A%val), A%val, A%ia, A%ja, 1.0_c_double, u, 0.0_c_double, y)
call check(maxval(abs(y - b)) < 1.0e-14_c_double, "csr_mv: y = A x")
y = 1
call csr_mv(A%n, A%n, size(A%val), A%val, A%ia, A%ja, 2.0_c_double, u, -0.5_c_double, y)
call check(maxval(abs(y - (2*b - 0.5_c_double))) < 1.0e-14_c_double, "csr_mv: beta, alpha")

! The callback without a context: the identity, so x = b
x = 0
status = solve_sparse(n, n, identity, c_null_ptr, b, x, res_iter=iter)
call check(status == SOLVER_SUCCESS, "mf identity: status")
call check(maxval(abs(x - b)) < 1.0e-12_c_double, "mf identity: solution")

if (failures > 0) then
    print '(a,i0,a)', "test_solver_fortran: ", failures, " failure(s)"
    error stop 1
end if
print '(a)', "test_solver_fortran: all tests passed"

end program

File: test/test_solver.cpp
// Tests for rbf_solver.h: the C interface over Eigen's iterative
// solvers, on the five-point Laplacian of a square grid with Dirichlet
// boundary, which is symmetric positive definite, and on the same with
// an upwind convection term, which is not. The right-hand side is
// manufactured from a known solution, so every method and
// preconditioner is checked against it; the matrix-free path runs the
// same systems through a callback and must land on the same answers.
//
// Build and run via CMake, from the repository root:
//
//   cmake -B build -DGPU_ASSEMBLY_ENABLE_SOLVER=ON
//   cmake --build build
//   ctest --test-dir build

#include <cmath>
#include <cstddef>
#include <cstdio>
#include <vector>

#include "rbf_solver.h"

#include "check.h"

// A sparse matrix in CSR, 0-based, as the solver takes it.
struct Csr {
    int n = 0;
    std::vector<int> ia, ja;
    std::vector<double> val;

    int nnz() const { return static_cast<int>(ja.size()); }

    // y += alpha * A * x
    void apply(double alpha, const double* x, double* y) const {
        for (int i = 0; i < n; ++i) {
            double s = 0;
            for (int p = ia[i]; p < ia[i + 1]; ++p)
                s += val[p] * x[ja[p]];
            y[i] += alpha * s;
        }
    }
};

// The five-point Laplacian on an m-by-m grid, -u_xx - u_yy discretized
// with the grid spacing folded in, plus an upwind discretization of
// c u_x when c is nonzero. Diagonally dominant either way.
static Csr laplacian(int m, double c = 0) {
    Csr A;
    A.n = m * m;
    A.ia.push_back(0);
    for (int j = 0; j < m; ++j) {
        for (int i = 0; i < m; ++i) {
            const int row = i + j * m;
            // sorted by column, which is not required but keeps the
            // matrix in the shape the assembly produces
            if (j > 0) {
                A.ja.push_back(row - m);
                A.val.push_back(-1);
            }
            if (i > 0) {
                A.ja.push_back(row - 1);
                A.val.push_back(-1 - c);
            }
            A.ja.push_back(row);
            A.val.push_back(4 + c);
            if (i < m - 1) {
                A.ja.push_back(row + 1);
                A.val.push_back(-1);
            }
            if (j < m - 1) {
                A.ja.push_back(row + m);
                A.val.push_back(-1);
            }
            A.ia.push_back(A.nnz());
        }
    }
    return A;
}

// A smooth field on the grid, the solution the right-hand side is made from.
static std::vector<double> exact(int m) {
    std::vector<double> u(static_cast<std::size_t>(m) * m);
    for (int j = 0; j < m; ++j)
        for (int i = 0; i < m; ++i)
            u[i + j * m] = std::sin(0.3 * i) * std::cos(0.2 * j) + 0.1 * i;
    return u;
}

static double max_abs_diff(const std::vector<double>& a, const std::vector<double>& b) {
    double d = 0;
    for (std::size_t i = 0; i < a.size(); ++i)
        d = std::fmax(d, std::fabs(a[i] - b[i]));
    return d;
}

static void callback(int nr, int nc, double alpha, const double* x, double* y, void* data) {
    const auto& A = *static_cast<const Csr*>(data);
    CHECK(nr == A.n && nc == A.n);
    A.apply(alpha, x, y);
}

// Every method and preconditioner on one matrix, against its known
// solution, from a zero initial guess.
static void test_csr(const Csr& A, const std::vector<double>& u, solver_method method) {
    std::vector<double> b(u.size(), 0);
    A.apply(1, u.data(), b.data());

    for (const auto precond : {PRECOND_NONE, PRECOND_JACOBI, PRECOND_ILU}) {
        std::vector<double> x(u.size(), 0);
        double err = -1;
        int iter = -1;
        const double tol = 1e-12;
        const int status =
            rbf_solve_csr_dp(A.n, A.nnz(), A.val.data(), A.ia.data(), A.ja.data(), b.data(),
                             x.data(), &err, &iter, &method, &precond, nullptr, &tol);
        CHECK(status == SOLVER_SUCCESS);
        CHECK(err >= 0 && err <= tol);
        CHECK(iter >= 0 && iter <= 2 * A.n);
        CHECK(max_abs_diff(x, u) < 1e-9);
    }
}

// The matrix-free solver on the same system, unpreconditioned.
static void test_matrix_free(const Csr& A, const std::vector<double>& u, solver_method method) {
    std::vector<double> b(u.size(), 0);
    A.apply(1, u.data(), b.data());

    std::vector<double> x(u.size(), 0);
    double err = -1;
    int iter = -1;
    const double tol = 1e-12;
    const int status = rbf_solve_mf_dp(A.n, A.n, callback, const_cast<Csr*>(&A), b.data(), x.data(),
                                       &err, &iter, &method, nullptr, &tol);
    CHECK(status == SOLVER_SUCCESS);
    CHECK(err >= 0 && err <= tol);
    CHECK(iter >= 0 && iter <= 2 * A.n);
    CHECK(max_abs_diff(x, u) < 1e-9);
}

static void test_defaults_and_guess() {
    const int m = 12;
    const Csr A = laplacian(m);
    const std::vector<double> u = exact(m);
    std::vector<double> b(u.size(), 0);
    A.apply(1, u.data(), b.data());

    // every optional argument left out: BiCGSTAB, Jacobi, 2n
    // iterations, epsilon
    std::vector<double> x(u.size(), 0);
    CHECK(rbf_solve_csr_dp(A.n, A.nnz(), A.val.data(), A.ia.data(), A.ja.data(), b.data(), x.data(),
                           nullptr, nullptr, nullptr, nullptr, nullptr, nullptr) == SOLVER_SUCCESS);
    CHECK(max_abs_diff(x, u) < 1e-9);

    // the exact solution as initial guess needs no iteration at all
    x = u;
    int iter = -1;
    CHECK(rbf_solve_csr_dp(A.n, A.nnz(), A.val.data(), A.ia.data(), A.ja.data(), b.data(), x.data(),
                           nullptr, &iter, nullptr, nullptr, nullptr, nullptr) == SOLVER_SUCCESS);
    CHECK(iter == 0);
    CHECK(max_abs_diff(x, u) == 0);
}

// The product on its own, against the reference loop, with y not read
// when beta is zero.
static void test_csr_mv() {
    const int m = 9;
    const Csr A = laplacian(m, 0.25);
    const std::vector<double> u = exact(m);
    std::vector<double> ref(u.size(), 0);
    A.apply(1, u.data(), ref.data());

    std::vector<double> y(u.size(), std::nan(""));
    rbf_csr_mv_dp(A.n, A.n, A.nnz(), A.val.data(), A.ia.data(), A.ja.data(), 1, u.data(), 0,
                  y.data());
    CHECK(max_abs_diff(y, ref) < 1e-14);

    // y = -0.5 y + 2 A u
    for (std::size_t i = 0; i < y.size(); ++i)
        y[i] = 0.1 * static_cast<double>(i);
    std::vector<double> expect(u.size());
    for (std::size_t i = 0; i < y.size(); ++i)
        expect[i] = -0.5 * y[i] + 2 * ref[i];
    rbf_csr_mv_dp(A.n, A.n, A.nnz(), A.val.data(), A.ia.data(), A.ja.data(), 2, u.data(), -0.5,
                  y.data());
    CHECK(max_abs_diff(y, expect) < 1e-14);
}

static void test_failures() {
    const int m = 20;
    const Csr A = laplacian(m);
    const std::vector<double> u = exact(m);
    std::vector<double> b(u.size(), 0);
    A.apply(1, u.data(), b.data());
    std::vector<double> x(u.size(), 0);

    // an iteration limit too low to converge: NO_CONVERGENCE, with the
    // last iterate and its residual reported
    const int one = 1;
    const double tol = 1e-14;
    double err = -1;
    int iter = -1;
    CHECK(rbf_solve_csr_dp(A.n, A.nnz(), A.val.data(), A.ia.data(), A.ja.data(), b.data(), x.data(),
                           &err, &iter, nullptr, nullptr, &one, &tol) == SOLVER_NO_CONVERGENCE);
    CHECK(iter == 1);
    CHECK(err > tol);

    // the same through the callback
    CHECK(rbf_solve_mf_dp(A.n, A.n, callback, const_cast<Csr*>(&A), b.data(), x.data(), &err, &iter,
                          nullptr, &one, &tol) == SOLVER_NO_CONVERGENCE);
    CHECK(iter == 1);

    // INVALID_INPUT: a null array, a row pointer that disagrees with
    // nnz, an enumerator outside the enum, a rectangular operator
    CHECK(rbf_solve_csr_dp(A.n, A.nnz(), nullptr, A.ia.data(), A.ja.data(), b.data(), x.data(),
                           nullptr, nullptr, nullptr, nullptr, nullptr,
                           nullptr) == SOLVER_INVALID_INPUT);
    CHECK(rbf_solve_csr_dp(A.n, A.nnz() - 1, A.val.data(), A.ia.data(), A.ja.data(), b.data(),
                           x.data(), nullptr, nullptr, nullptr, nullptr, nullptr,
                           nullptr) == SOLVER_INVALID_INPUT);
    const auto bad_method = static_cast<solver_method>(7);
    CHECK(rbf_solve_csr_dp(A.n, A.nnz(), A.val.data(), A.ia.data(), A.ja.data(), b.data(), x.data(),
                           nullptr, nullptr, &bad_method, nullptr, nullptr,
                           nullptr) == SOLVER_INVALID_INPUT);
    const auto bad_precond = static_cast<solver_precond>(7);
    CHECK(rbf_solve_csr_dp(A.n, A.nnz(), A.val.data(), A.ia.data(), A.ja.data(), b.data(), x.data(),
                           nullptr, nullptr, nullptr, &bad_precond, nullptr,
                           nullptr) == SOLVER_INVALID_INPUT);
    CHECK(rbf_solve_mf_dp(A.n, A.n + 1, callback, const_cast<Csr*>(&A), b.data(), x.data(), nullptr,
                          nullptr, nullptr, nullptr, nullptr) == SOLVER_INVALID_INPUT);
    CHECK(rbf_solve_mf_dp(A.n, A.n, nullptr, nullptr, b.data(), x.data(), nullptr, nullptr, nullptr,
                          nullptr, nullptr) == SOLVER_INVALID_INPUT);
}

int main() {
    const int m = 20;
    const std::vector<double> u = exact(m);

    // symmetric positive definite: both methods
    const Csr spd = laplacian(m);
    test_csr(spd, u, SOLVER_CG);
    test_csr(spd, u, SOLVER_BICGSTAB);
    test_matrix_free(spd, u, SOLVER_CG);
    test_matrix_free(spd, u, SOLVER_BICGSTAB);

    // nonsymmetric: BiCGSTAB only
    const Csr nonsym = laplacian(m, 0.5);
    test_csr(nonsym, u, SOLVER_BICGSTAB);
    test_matrix_free(nonsym, u, SOLVER_BICGSTAB);

    test_defaults_and_guess();
    test_csr_mv();
    test_failures();

    return report("test_solver");
}
